File: loadmore/src/main/java/com/tuuzed/androidx/list/loadmore/LoadMore.java
package com.tuuzed.androidx.list.loadmore;

import android.util.Log;
import android.view.View;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

@SuppressWarnings("unused")
public class LoadMore {
    private static final String TAG = "LoadMore";
    private RecyclerView.Adapter originalAdapter;
    private OnLoadMoreListener onLoadMoreListener;
    private LoadMoreController loadMoreController;

    private View footerView;
    private int footerResId = View.NO_ID;

    private View noMoreView;
    private int noMoreResId = View.NO_ID;

    private View loadFailedView;
    private int loadFailedResId = View.NO_ID;

    @NonNull
    public static LoadMore of(RecyclerView.Adapter originalAdapter) {
        return new LoadMore(originalAdapter);
    }

    public LoadMore(@NonNull RecyclerView.Adapter originalAdapter) {
        this.originalAdapter = originalAdapter;
    }

    @NonNull
    public LoadMore setOnLoadMoreListener(@NonNull OnLoadMoreListener listener) {
        this.onLoadMoreListener = listener;
        return this;
    }

    @NonNull
    public LoadMore setLoadMoreController(@NonNull LoadMoreController loadMoreController) {
        this.loadMoreController = loadMoreController;
        return this;
    }

    @NonNull
    public LoadMore setFooterView(View footerView) {
        this.footerView = footerView;
        return this;
    }

    @NonNull
    public LoadMore setFooterResId(@LayoutRes int footerResId) {
        this.footerResId = footerResId;
        return this;
    }

    @NonNull
    public LoadMore setNoMoreView(View noMoreView) {
        this.noMoreView = noMoreView;
        return this;
    }

    @NonNull
    public LoadMore setNoMoreResId(@LayoutRes int noMoreResId) {
        this.noMoreResId = noMoreResId;
        return this;
    }

    @NonNull
    public LoadMore setLoadFailedView(View loadFailedView) {
        this.loadFailedView = loadFailedView;
        return this;
    }

    @NonNull
    public LoadMore setLoadFailedResId(@LayoutRes int loadFailedResId) {
        this.loadFailedResId = loadFailedResId;
        return this;
    }

    public void into(@NonNull RecyclerView recyclerView) {
        if (loadMoreController == null) {
            loadMoreController = new LoadMoreController();
        }
        if (onLoadMoreListener == null) {
            onLoadMoreListener = new OnLoadMoreListener() {
                @Override
                public void onLoadMore(@NonNull LoadMoreController controller) {
                    Log.w(TAG, "未设置OnLoadMoreListener");
                }
            };
        }
        LoadMoreAdapter loadMoreAdapter = new LoadMoreAdapter(
                originalAdapter, onLoadMoreListener, loadMoreController,
                footerView, footerResId,
                noMoreView, noMoreResId,
                loadFailedView, loadFailedResId
        );
        loadMoreAdapter.setHasStableIds(originalAdapter.hasStableIds());
        recyclerView.setAdapter(loadMoreAdapter);
    }
}

File: loadmore/src/main/java/com/tuuzed/androidx/recyclerview/adapter/loadmore/LoadMore.kt
package com.tuuzed.androidx.recyclerview.adapter.loadmore

import android.view.View
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView
import com.tuuzed.androidx.recyclerview.adapter.RecyclerViewAdapter

@JvmOverloads
fun RecyclerViewAdapter.withLoadMore(
        recyclerView: RecyclerView,
        loadMoreState: LoadMoreState = LoadMoreState(),
        enableLoadMore: Boolean = true,
        enableShowNoMore: Boolean = true,
        footerView: View? = null,
        noMoreView: View? = null,
        loadFailedView: View? = null,
        @LayoutRes footerViewLayoutRes: Int = View.NO_ID,
        @LayoutRes noMoreViewLayoutRes: Int = View.NO_ID,
        @LayoutRes loadFailedViewLayoutRes: Int = View.NO_ID,
        listener: OnLoadMoreListener
): RecyclerViewAdapter {
    LoadMore.with(
            recyclerView,
            loadMoreState,
            enableLoadMore,
            enableShowNoMore,
            footerView, noMoreView,
            loadFailedView,
            footerViewLayoutRes,
            noMoreViewLayoutRes,
            loadFailedViewLayoutRes,
            listener
    )
    return this
}

object LoadMore {

    @JvmOverloads
    fun with(recyclerView: RecyclerView,
             loadMoreState: LoadMoreState = LoadMoreState(),
             enableLoadMore: Boolean = true,
             enableShowNoMore: Boolean = true,
             footerView: View? = null,
             noMoreView: View? = null,
             loadFailedView: View? = null,
             @LayoutRes footerViewLayoutRes: Int = View.NO_ID,
             @LayoutRes noMoreViewLayoutRes: Int = View.NO_ID,
             @LayoutRes loadFailedViewLayoutRes: Int = View.NO_ID,
             listener: OnLoadMoreListener
    ) {
        val adapter = recyclerView.adapter
                ?: throw NullPointerException("recyclerView.adapter == null")
        val loadMoreAdapter = LoadMoreAdapter(
                originalAdapter = adapter,
                loadMoreState = loadMoreState,
                enableShowNoMore = enableShowNoMore,
                footerView = footerView,
                noMoreView = noMoreView,
                loadFailedView = loadFailedView,
                footerViewLayoutRes = footerViewLayoutRes,
                noMoreViewLayoutRes = noMoreViewLayoutRes,
                loadFailedViewLayoutRes = loadFailedViewLayoutRes,
                onLoadMoreListener = listener
        ).also {
            it.setHasStableIds(it.originalAdapter.hasStableIds())
        }
        loadMoreAdapter.loadMoreEnabled = enableLoadMore
        recyclerView.adapter = loadMoreAdapter
    }


}
